lib/datadog/appsec/processor.rb

```diff
@@ -0,0 +1,113 @@
+# typed: ignore
+
+require 'datadog/appsec/assets'
+
+module Datadog
+  module AppSec
+    # Processor integrates libddwaf into datadog/appsec
+    class Processor
+      # Interface object to check using case .. when
+      module IOLike
+        def read; end
+        def rewind; end
+
+        def self.===(other)
+          instance_methods.reduce(true) { |ok, meth| ok && other.respond_to?(meth) }
+        end
+      end
+
+      def initialize
+        unless load_libddwaf && load_ruleset && create_waf_handle
+          Datadog.logger.warn { 'appsec is disabled, cause: errors' }
+        end
+      end
+
+      def ready?
+        !@ruleset.nil? && !@handle.nil?
+      end
+
+      def context
+        Datadog::AppSec::WAF::Context.new(@handle)
+      end
+
+      private
+
+      def load_libddwaf
+        Processor.require_libddwaf && Processor.libddwaf_required?
+      end
```

Comment:
IIRC there is none right now, but there should be: on unsupported platforms (a.k.a `ruby` gem platform) libddwaf attempts to load the shared library for that platform, which currently raises a `LoadError`. This is hacky at best though, and should be a proper interface.

My plan is to change libddwaf on `ruby` to either compile libddwaf from source (if build deps are met) or be a true shim, and in all cases you can query feature availability (which could be 'none'). In that case it could successfully be required (no `LoadError`) yet the needed features be unavailable. Since there is no such API currently I kept it as a simple constant check. The name is probably misleading though, it does not mean "required as Ruby understands it" but "required as in being able to provide the actual feature".

I did not want the refactoring to scope-creep though, hence why I kept the name and behaviour as is.

Reply:
👍  Seems reasonable. Perhaps leave a comment with these notes as well?

Reply:
Done